vcd/resource_vcd_snat.go

```diff
@@ -131,14 +141,14 @@ func resourceVcdSNATRead(d *schema.ResourceData, meta interface{}) error {
 	providedNetworkName := d.Get("network_name")
 	if nil != providedNetworkName && providedNetworkName.(string) != "" {
 		for _, r := range e.EdgeGateway.Configuration.EdgeGatewayServiceConfiguration.NatService.NatRule {
-			if r.RuleType == "SNAT" && r.GatewayNatRule.OriginalIP == d.Id() {
+			if r.RuleType == "SNAT" && r.GatewayNatRule.TranslatedIP == d.Get("internal_ip").(string) && r.GatewayNatRule.Interface.Name == d.Get("network_name").(string) {
 				found = true
-				d.Set("external_ip", r.GatewayNatRule.TranslatedIP)
+				d.Set("external_ip", r.GatewayNatRule.OriginalIP)
 			}
 		}
-	} else {
+	} else { // TODO remove after network_name becomes mandatory
 		for _, r := range e.EdgeGateway.Configuration.EdgeGatewayServiceConfiguration.NatService.NatRule {
-			if r.RuleType == "SNAT" && r.GatewayNatRule.OriginalIP == d.Id() && r.GatewayNatRule.Interface.Name == d.Get("network_name").(string) {
+			if r.RuleType == "SNAT" && r.GatewayNatRule.OriginalIP == d.Get("internal_ip").(string) {
```

Comment:
Do I read correctly that this is comparing original IP and if it matches it considers that this is our instance of the rule? What would happen if the user created a rule manually on same original IP with different ports? Why is it ignoring original port, translated IP and translated port?

Overall, using field values seems error prone, why not to just base on the rule ID? Then again I may be missing the context here.

Would like @Didainius review on this part as well.

Reply:
I think this was always the way until now. I saw this in govcd and didn't like it too much as well. I think there's a TODO to make govcd part being able to look up and modify by IDs. Then we could tune provider and store them in the statefile.

Reply:
I don't know why they haven't used ID's. I see according `types` ID should be accessible. I investigate that

Reply:
Thanks! ID would be much more robust and less risky.

Reply:
Adding `terraform.tfstate` snippet for reference (to ease the discussion) where one can see the issue with current "id" implementation:

```
                "vcd_snat.outbound": {
                    "type": "vcd_snat",
                    "depends_on": [],
                    "primary": {
                        "id": "192.168.0.0/24",
                        "attributes": {
                            "edge_gateway": "EdgeToVMs",
                            "external_ip": "10.150.211.102/32",
                            "id": "192.168.0.0/24",
                            "internal_ip": "192.168.0.0/24"
                        },
                        "meta": {},
                        "tainted": false
                    },
                    "deposed": [],
                    "provider": "provider.vcd"
                },
```

Compare that to the `vcd_dnat` rule, which, though doesn't use the real ID either, has rendered both IPs and ports there for uniqueness (BTW, "\u003e" is the `>` sign):

```
                "vcd_dnat.ssh_vm11": {
                    "type": "vcd_dnat",
                    "depends_on": [],
                    "primary": {
                        "id": "10.150.211.102/32:2211 \u003e 192.168.0.11/32:22",
                        "attributes": {
                            "edge_gateway": "EdgeToVMs",
                            "external_ip": "10.150.211.102/32",
                            "icmp_sub_type": "any",
                            "id": "10.150.211.102/32:2211 \u003e 192.168.0.11/32:22",
                            "internal_ip": "192.168.0.11/32",
                            "port": "2211",
                            "protocol": "TCP",
                            "translated_port": "22"
                        },
                        "meta": {},
                        "tainted": false
                    },
                    "deposed": [],
                    "provider": "provider.vcd"
                }
```

Reply:
From a more global perspective, when checking all my state files, it's only SNAT and DNAT rules that use "arbitrary" values for "id" field. Other resources use names nicely:

```
$ grep "\"id" 2.2.0-features/terraform.tfstate
                        "id": "10.150.211.102/32:2211 \u003e 192.168.0.11/32:22",
                            "id": "10.150.211.102/32:2211 \u003e 192.168.0.11/32:22",
                        "id": "EdgeToVMs",
                            "id": "EdgeToVMs",
                        "id": "TfNet",
                            "id": "TfNet",
                        "id": "192.168.0.0/24",
                            "id": "192.168.0.0/24",
                        "id": "TfVApp",
                            "id": "TfVApp",
                        "id": "TfVAppNet",
                            "id": "TfVAppNet",
                        "id": "TerraformVM11",
                            "id": "TerraformVM11",
$ grep "\"id" vdc/terraform.tfstate
                        "id": "TfAllocationVDC",
                            "id": "TfAllocationVDC",
                        "id": "TfPayVDC",
                            "id": "TfPayVDC",
                        "id": "TfReservationVDC",
                            "id": "TfReservationVDC",
$ grep "\"id" catalogs/terraform.tfstate
                        "id": "OperatingSystems",
                            "id": "OperatingSystems",
                        "id": "OperatingSystems:Linux",
                            "id": "OperatingSystems:Linux",
                        "id": "OperatingSystems:dsl-4.11.rc2.iso",
                            "id": "OperatingSystems:dsl-4.11.rc2.iso",
lvirbalas-a01:configurations lvirbalas$ grep "\"id" demo/terraform.tfstate
                        "id": "10.150.211.101/32:2227 \u003e 192.168.0.7/32:22",
                            "id": "10.150.211.101/32:2227 \u003e 192.168.0.7/32:22",
                        "id": "EdgeToVMware",
                            "id": "EdgeToVMware",
                        "id": "tf-disk",
                            "id": "tf-disk",
                        "id": "TerraformVApp_TerraformVM7_dsl-4.11.rc2.iso",
                            "id": "TerraformVApp_TerraformVM7_dsl-4.11.rc2.iso",
                        "id": "TerraformNetwork",
                            "id": "TerraformNetwork",
                        "id": "192.168.0.0/24",
                            "id": "192.168.0.0/24",
                        "id": "TerraformVApp",
                            "id": "TerraformVApp",
                        "id": "TerraformVAppNet",
                            "id": "TerraformVAppNet",
                        "id": "TerraformVM7",
                            "id": "TerraformVM7",
                        "id": "TerraformVM9",
                            "id": "TerraformVM9",
```

Reply:
We had to figure out regardind ID - it isn't return on creation. Postponed at the moment.